text/0000-route-data-loading.md

```diff
@@ -0,0 +1,69 @@
+- Start Date: 2019-07-01
+- Relevant Team(s): Ember.js
+- RFC PR: 
+- Tracking: 
+
+# Route Data Loading for Components
+
+## Summary
+
+[#499 An alternative to Controllers](https://github.com/emberjs/rfcs/pull/499) proposes an API for rendering a component from a route. This RFC builds on it by proposing an API to pass multiple arguments to a component from a route. It's based off the [design proposed by @chadhietala](https://gist.github.com/chadhietala/50b977a7d3476069892d351c65af418c) and much of the text is lifted from there.
+
+## Motivation
+
+The `model()` hook on routes is confusing, because it implies that only one piece of data will be fetched, and it uses the term "model" that is more typically associated with backend development. We should have a better name, and an API that allows data to be passed from a route into a component as multiple named arguments.
+
+## Detailed design
+
+This RFC proposes adding a `load()` hook to `Route`. The route is in the `loading` state while this hook is run. The keys of the object returned from `load()` become the named arguments to the component.
```

Comment:
While I like the approach, I see a few things that are underspecified or deserve to be defined more explicitly:
* It will continue to wait for returned promises as `model()` does, affecting the `loading` state as well as the rendering of the component, as proposed by RFC499. "while this hook is run" is not very precise in this regard.
* I assume the arguments will be the same as for `model()`?
* Will `load()` always be called, even when the "model" is supplied in the `link-to`. I.e. how does it relate to https://github.com/emberjs/rfcs/pull/283
* The `model` hook has a default implementation that will make loading work magically in simple scenarios without actually implementing the hook, see http://api.emberjs.com/ember/release/classes/Route/methods/model?anchor=model. Do we want to copy over this (too) magic behaviour? I think the new `load()` API proposed here would be a good way to get rid of this, and force users to make everything explicit!
* The "keys" are what `Object.keys()` would return, i.e. only the own (`hasOwnProperty`) props, and nothing from prototypes, right?
* What happens when the returned (resolved) value is not an object? (a primitive value, or an array) How is the route component called in this case? Or is the error component rendered? Or does it simply throw an exception?

Reply:
> It will continue to wait for returned promises as model() does, affecting the loading state as well as the rendering of the component, as proposed by RFC499. "while this hook is run" is not very precise in this regard.

That's the intent. I will change this to "The route is in the loading state while this hook is run. When this hook has finished running and any promises it returns have resolved, the route will change to the resolved state."

> I assume the arguments will be the same as for model()?

Yes, good clarification. I will add that.

> Will load() always be called, even when the "model" is supplied in the link-to. I.e. how does it relate to #283

I wasn't previously familiar with #283, but I think the proposal is good. Maybe this is a good opportunity to achieve the goals of #283 without introducing a new base class and the potentially-confusingly-similar `findModel` hook. I will update this to specify that the `load` hook will always be run, and suggest that this subsume #283. As a follow-up to this RFC, we also might want to reconsider the word "model" for `link-to`...

> The model hook has a default implementation that will make loading work magically in simple scenarios without actually implementing the hook, see http://api.emberjs.com/ember/release/classes/Route/methods/model?anchor=model. Do we want to copy over this (too) magic behaviour? I think the new load() API proposed here would be a good way to get rid of this, and force users to make everything explicit!

Agreed - I will specify that `load()` will not support the "magic" behavior.

> The "keys" are what Object.keys() would return, i.e. only the own (hasOwnProperty) props, and nothing from prototypes, right?

Righto. I will be more explicit about this.

> What happens when the returned (resolved) value is not an object? (a primitive value, or an array) How is the route component called in this case? Or is the error component rendered? Or does it simply throw an exception?

I think we throw an exception. The `error` state is for errors with the data that is being fetched like being unauthorized, but in this case, the error is a mistake in the front-end code.